File: src/main/webapp/scripts/models/appuser.js
//***************************************************************************************
//* AppUser
//***************************************************************************************

function AppUser(bAuthenticated, iAppUserId, sFirstName, sLastName, sUsername, sPasscode, dtSignUp)
{
	this.mbChanged			=	undefined;
	this.msFirstName 		= 	Common.defaultIfUndefined(sFirstName, "");
	this.miAppUserId		=	Common.defaultIfUndefined(iAppUserId, 0);
	this.msLastName 		= 	Common.defaultIfUndefined(sLastName, "");
	this.mObjLoginInfo		=	undefined;
	this.mArrObjObservers	=	[];
	this.mObjOffice			=	undefined;
	this.mDtSignUp 			=	dtSignUp;
}

AppUser.prototype.addObserver		=	function(objObserver)
{
	if(objObserver != undefined)
		{
		if(this.mArrObjObservers == undefined)
			this.mArrObjObservers 	=	[];
		
		this.mArrObjObservers.push(objObserver);
		}
	
	return(this);
};

AppUser.copyFromJSON				=	function(objJSON, objAppUser)
{
	if(objAppUser == undefined)
		objAppUser = new AppUser();
	
	for(prop in objJSON)
		switch(prop)
			{
			case 'office':
				objAppUser.setOffice(Office.copyFromJSON(undefined, eval("objJSON." + prop), objAppUser.getOffice()));
				break;
			case 'loginInfo':
				objAppUser.setLoginInfo(LoginInfo.copyFromJSON(eval("objJSON." + prop), objAppUser.getLoginInfo()));
				break;
			default:
				objAppUser["set" + prop.substring(0, 1).toUpperCase() + prop.substring(1)].call(objAppUser, eval("objJSON." + prop));
			}
	
	return(objAppUser);
};

AppUser.prototype.getAppUserId		=	function()
{
	if(this.miAppUserId == undefined)
		this.miAppUserId = 0;
	
	return(this.miAppUserId);
};

AppUser.prototype.getFirstName		=	function()
{
	if(this.msFirstName == undefined)
		this.msFirstName = "";
	
	return(this.msFirstName);
};

AppUser.prototype.getLastName		=	function()
{
	if(this.msLastName == undefined)
		this.msLastName = "";
	
	return(this.msLastName);
};

AppUser.prototype.getLoginInfo		=	function()
{
	if(this.mObjLoginInfo == undefined)
		this.mObjLoginInfo = new LoginInfo();
	
	return(this.mObjLoginInfo);
};

AppUser.prototype.getOffice			=	function()
{
	if(this.mObjOffice == undefined)
		this.mObjOffice = new Office();
	
	return(this.mObjOffice);
};

AppUser.prototype.getSignupDate		=	function()
{
	return(this.mDtSignUp);
};

AppUser.prototype.hasChanged		=	function()
{
	if(this.mbChanged == undefined)
		this.mbChanged = false;
	
	return(this.mbChanged);
};

AppUser.prototype.notifyObservers	=	function(objArgument)
{
	this.mbChanged	=	true;
	
	for(var iLength = this.mArrObjObservers.length, iLoop = 0; iLoop < iLength; iLoop++)
		this.mArrObjObservers[iLoop].call(this, objArgument);
	
	this.mbChanged	=	false;
};

AppUser.prototype.packageToJSON		=	function()
{
	return(	{	appUserId:			this.getAppUserId()
			,	firstName:			this.getFirstName()
			,	lastName:			this.getLastName()
			,	loginInfo:			this.getLoginInfo().packageToJSON()
			,	office:				this.getOffice().packageToJSON()
			,	signupDate:			this.getSignupDate()
			});
};

AppUser.prototype.setAppUserId		=	function(iAppUserId)
{
	if(iAppUserId == undefined)
		this.miAppUserId			=	0;
	else
		this.miAppUserId			=	parseInt(iAppUserId);
	
	return(this);
};

AppUser.prototype.setFirstName		=	function(sFirstName)
{
	if(sFirstName == undefined)
		this.msFirstName			=	"";
	else
		this.msFirstName 			=	sFirstName;
	
	return(this);
};

AppUser.prototype.setLastName		=	function(sLastName)
{
	if(sLastName == undefined)
		this.msLastName				=	"";
	else
		this.msLastName 			=	sLastName;
	
	return(this);
};

AppUser.prototype.setLoginInfo		=	function(objLoginInfo)
{
	if(objLoginInfo == undefined)
		this.mObjLoginInfo 			=	new LoginInfo();
	else
		this.mObjLoginInfo			=	objLoginInfo;
	
	return(this);
};

AppUser.prototype.setOffice			=	function(objOffice)
{
	if(objOffice == undefined)
		this.mObjOffice				=	new Office();
	else
		this.mObjOffice 			=	objOffice;
	
	return(this);
};

AppUser.prototype.setSignupDate		=	function(dtSignUp)
{
	this.mDtSignUp 					=	dtSignUp;
	
	return(this);
};

File: src/main/java/net/draconia/askaround/domain/AppUser.java
package net.draconia.askaround.domain;

import java.io.Serializable;

import java.util.Date;

import net.draconia.askaround.util.DateSerializer;

import org.codehaus.jackson.map.annotate.JsonSerialize;

public class AppUser implements Serializable
{
	private static final long serialVersionUID = 6289188460713227245L;
	
	private Date mDtSignup;
	private Integer miAppUserId;
	private LoginInfo mObjLoginInfo;
	private Office mObjOffice;
	private String msFirstName, msLastName;

	public AppUser()
	{ }
	
	public AppUser(final Integer iAppUserId, final String sFirstName, final String sLastName, final Date dtSignup, final Office objOffice)
	{
		setAppUserId(iAppUserId);
		setFirstName(sFirstName);
		setLastName(sLastName);
		setOffice(objOffice);
		setSignupDate(dtSignup);
	}
	
	public Integer getAppUserId()
	{
		if(miAppUserId == null)
			miAppUserId = 0;
		
		return(miAppUserId);
	}
	
	public String getFirstName()
	{
		if(msFirstName == null)
			msFirstName = "";
		
		return(msFirstName);
	}
	
	public String getLastName()
	{
		if(msLastName == null)
			msLastName = "";
		
		return(msLastName);
	}
	
	public LoginInfo getLoginInfo()
	{
		if(mObjLoginInfo == null)
			mObjLoginInfo = new LoginInfo();
		
		return(mObjLoginInfo);
	}
	
	public Office getOffice()
	{
		if(mObjOffice == null)
			mObjOffice = new Office();
		
		return(mObjOffice);
	}
	
	/**
	 * getSignupDate - Returns the SignUp date for this AppUser
	 * @return Signup date in the AppUser record
	 */
	// The annotation permits the Jackson library to encode the date into the JSON
	@JsonSerialize(using = DateSerializer.class)
	public Date getSignupDate()
	{
		if(mDtSignup == null)
			mDtSignup = new Date();
		
		return(mDtSignup);
	}
	
	public void setAppUserId(final Integer iId)
	{
		if(iId == null)
			miAppUserId = 0;
		else
			miAppUserId = iId;
	}
	
	public void setFirstName(final String sFirstName)
	{
		if(sFirstName == null)
			msFirstName = "";
		else
			msFirstName = sFirstName;
	}
	
	public void setLastName(final String sLastName)
	{
		if(sLastName == null)
			msLastName = "";
		else
			msLastName = sLastName;
	}
	
	public void setLoginInfo(final LoginInfo objLoginInfo)
	{
		if(objLoginInfo == null)
			mObjLoginInfo = new LoginInfo();
		else
			mObjLoginInfo = objLoginInfo;
	}
	
	public void setOffice(final Office objOffice)
	{
		if(objOffice == null)
			mObjOffice = new Office();
		else
			mObjOffice = objOffice;
	}
	
	public void setSignupDate(final Date dtSignup)
	{
		if(mDtSignup == null)
			mDtSignup = new Date();
		else
			mDtSignup = dtSignup;
	}
}